### ISSUE TITLE
[seo-magic] Uncatched HttpClientException in getImages() Method

### ISSUE THREAD

**Issue by bnjmnhssnn** [getgrav/grav-premium-issues#108]
I get this exception when I try to regenerate all SEO data:
```
[2021-06-11 06:56:11] grav.CRITICAL: HTTP/1.1 404 Not Found returned for "http://my.site.de/...
Trace: 
#0 /var/www/web/user/plugins/seo-magic/vendor/symfony/http-client/Response/ResponseTrait.php(308): Symfony\Component\HttpClient\Response\CurlResponse->checkStatusCode() 
#1 /var/www/web/user/plugins/seo-magic/vendor/symfony/http-client/Response/CurlResponse.php(234): Symfony\Component\HttpClient\Response\CurlResponse->doDestruct() 
#2 /var/www/web/user/plugins/seo-magic/classes/SEOGenerator.php(268): Symfony\Component\HttpClient\Response\CurlResponse->__destruct() 
#3 /var/www/web/user/plugins/seo-magic/vendor/symfony/dom-crawler/Crawler.php(355): Grav\Plugin\SEOMagic\SEOGenerator::Grav\Plugin\SEOMagic\{closure}(Object(Symfony\Component\DomCrawler\Crawler), 4) 
[...]
```
This exception cannot be handled here with `try catch`. It is unclear to me where and when the destructor is being called, maybe it has to do with the async behavior of the Symfony HttpClient? However, I was able to fix this quick and dirty by forcing and catching the Exception directly after the line `$image_responses[$src] = $client->request('HEAD', $src);`  with following code:
```php

// https://symfony.com/doc/current/http_client.html#handling-exceptions
// Force exception with getHeaders(), getContent() or toArray()
try {
   $image_responses[$src]->getHeaders();
} catch (\Exception $e) { /* do nothing */}
```
The drawback is that the code is blocking, now. Later, my colleague refactored the method and now is works without `try catch`, but we don't understand the exact mechanism behind this thing:
```php
public static function getImages(Crawler $crawler, $client, $host_base)
{
    $image_elements = $crawler->filterXPath('//img');
    $images = [];

    if ($image_elements->count() > 0) {
        $image_elements->each(function (Crawler $node) use (&$client, $host_base, &$images, &$image_responses) {
            $src = trim($node->attr('src'));
            $alt = trim($node->attr('alt'));

            if ($src) {
                $src = static::getValidLink($src, $host_base);
                $response = $client->request('HEAD', $src);
                $images[$src] = [
                    'alt' => $alt, 
                    'external' => static::isExternal($src, $host_base),
                    'statusCode' => $response->getStatusCode(),
                ];
            }
        });
    }
    return $images;
}

**Comment by rhukster** [comment]
This is due to how the Symfony HTTPClient works.  The initial `request()` method is asynchronous and does not wait for a response.  That's why I make those calls and store them in an array.  Then after I loop over those requests and check the status codes.  

Using `request('HEAD'...` is lighter-weight and fast as it doesn't actually download the full payload, just the response headers as all we care about is if the status is 404 or not.  

I will check to see why a 404 error is throwing an exception though, it shouldn't be doing that.  Keep using your workaround for now, but with it, you are losing some performance.

**Comment by w00fz** [comment]
The exception is thrown by design with symfony httpclient. By default they throw for any non `2XX`.
 It can however be prevented by just passing `false` as argument for the methods `getContent()`,  `toArray()`, `getHeaders()`. In our case since it’s something that we want to manually handle later, it should be `getHeaders(false)`.

**Comment by rhukster** [comment]
Can you please just try changing lline 273 to:

```php
$image_data['status'] = $image->getInfo('http_code');
```

This way we simply shortcut the `getStatusCode()` method throwing the exceptions.

**Comment by bnjmnhssnn** [comment]
>Can you please just try changing lline 273 to:
>`$image_data['status'] = $image->getInfo('http_code');`

No, this will not fix the error. Solution below:

1. To reproduce the error, you need 2 identical image sources on the same Grav page.
2. See the [Symfony HTTP Client docs about exeption handling](https://symfony.com/doc/current/http_client.html#handling-exceptions), and especially the part about the destruction-time behavior.
3. See the line marked with HERE in the first `getImages()` loop:
```php
if ($image_elements->count() > 0) {
    $image_elements->each(function (Crawler $node) use (&$client, $host_base, &$images, &$image_responses) {
        $src = trim($node->attr('src'));
        $alt = trim($node->attr('alt'));

        if ($src) {
            $src = static::getValidLink($src, $host_base);
            $images[$src] = ['alt' => $alt, 'external' => static::isExternal($src, $host_base)];
            $image_responses[$src] = $client->request('HEAD', $src); //<- HERE
        }
    });
}
```
When the result of `$client->request()` is assigned to an already existing array key, the array has no pointer to the previous response object any more. This doesn't call the destructor of the previous response, it will live on.

The second loop over the collected responses will initialize each of them by calling the `getStatusCode()` method, but will leave out the previously 'overwritten' one. This response remains uninitialized and will throw an exception on destruction time that you cannot catch inside this method. As @xabbuh stated, this is the expected behavior.

To fix the problem, just check if the array key exists and if yes, skip. This way, the requests will remain concurrent, contrary to the workarounds above in this thread.
```php
if ($src) {
    $src = static::getValidLink($src, $host_base);
    if(!isset($images[$src])) {
        $images[$src] = ['alt' => $alt, 'external' => static::isExternal($src, $host_base)];
        $image_responses[$src] = $client->request('HEAD', $src);
    }
}

**Comment by rhukster** [comment]
AH!!!! Now I see the issue.  I was testing with broken images and simply couldn't replicate the issue you were reporting.  They seemed to work fine. 

But now you mention "two" references to the same broken image and I can definitely see the issue you are describing and your solution works.  The only thing I want to check is if it's worth keeping track of how many of the same broken image (and the same goes for links) there are on the page, i.e/, if you have 10 references to the same broken image, with the current code, it would report only 1 broken image.

**Comment by bnjmnhssnn** [comment]
>But now you mention "two" references to the same broken image

I didn't notice that before, sorry.

**Comment by rhukster** [comment]
This should be fixed in release version 2.0.2.   I've gone with a change in the approach so it keeps all the broken image references as there could be a difference in ALT tag that might help identify which one is the problem.
